### PR TITLE
feat(debug): service cancelling and profiling build

### PR DESCRIPTION
**Pull request by Trantorian1**
# Pull Request type

- Feature

## What is the current behavior?

Each capability in Madara starts as a `Service`: currently there is no way for services to communicate shutdown intent between each other. That is to say for example the rpc service can't ask all the other services to gracefully shutdown, and can only manage its own life cycle.

## What is the new behavior?

This pr adds a shared `tokio_util::sync::CancellationToken` to each service in a `ServiceGroup`. This is directly passed to `graceful_shutdown` so that if any service requests a cancellation, other services will also enter shutdown mode.

Besides that, this pr adds a profiling build target for use in benchmarks and performance analysis. This mostly adds debug symbols to a release-equivalent build and uses `symbol-mangling` to demangle them for clearer perf reports.

> [!WARNING]
> This is _not_ an error handling mechanism. The idea is not to allow a service to gracefully shutdown related services in case of an error, which is why `tokio_util::sync::CancellationToken::child_token` is never used. This is only targeted at features which would allow the global halting of Madara from a local service (such as an admin rpc endpoints).

## Does this introduce a breaking change?

No.

### CODE REVIEW

File: crates/node/src/cli/sync.rs [hunk 1]
```diff
@@ -62,6 +62,13 @@ pub struct SyncParams {
     #[clap(env = "MADARA_N_BLOCKS_TO_SYNC", long, value_name = "NUMBER OF BLOCKS")]
     pub n_blocks_to_sync: Option<u64>,
 
+    /// Gracefully shutdown Madara once it has finished to synchronize all
```
Comment (shamsasari) [inline crates/node/src/cli/sync.rs]: Grammer: finished synchronizing all blocks.

File: crates/node/src/cli/sync.rs [hunk 2]
```diff
@@ -62,6 +62,13 @@ pub struct SyncParams {
...
+    /// blocks. This can either be once the node has caught up with the head of
+    /// the chain or when it has synced as many blocks as specified by
+    /// `n-blocks-to-sync`.
+    #[clap(env = "MADARA_STOP_ON_SYNC", long, value_name = "STOP ON SYNC", default_value_t = false)]
```
Comment (shamsasari) [inline crates/node/src/cli/sync.rs]: The value_name isn't used for bool flags - you can see that in the `--help` output.

File: crates/node/src/cli/sync.rs [hunk 3]
```diff
@@ -62,6 +62,13 @@ pub struct SyncParams {
     #[clap(env = "MADARA_N_BLOCKS_TO_SYNC", long, value_name = "NUMBER OF BLOCKS")]
     pub n_blocks_to_sync: Option<u64>,
 
+    /// Gracefully shutdown Madara once it has finished to synchronize all
+    /// blocks. This can either be once the node has caught up with the head of
+    /// the chain or when it has synced as many blocks as specified by
+    /// `n-blocks-to-sync`.
```
Comment (shamsasari) [inline crates/node/src/cli/sync.rs]: The backticks don't do anything and appear in the `--help` output, making it look a bit weird to the user. Change this to `--n-blocks-to-sync`

File: crates/node/src/service/block_production.rs [hunk 4]
```diff
@@ -104,7 +108,7 @@ impl Service for BlockProductionService {
             };
 
             // display devnet welcome message :)
-            // we display it to stdout instead of stderr
+            // we display itto stdout instead of stderr
```
Comment (shamsasari) [inline crates/node/src/service/block_production.rs]: revert

**Review by shamsasari — CHANGES_REQUESTED** [review]

**Review by shamsasari — APPROVED** [review]
LGTM

**Comment by antiyro** [comment]
lgtm but I would use `cancellation_token.is_cancelled()` instead of `CTRL_C` to make it cleaner and easier to maintain. If you receive `ctrl_c` and `sigterm` you could use cancelation token instead (completely removing `CTRL_C` if possible). Check https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.is_cancelled

**Comment by Trantorian1** [comment]
> lgtm but I would use `cancellation_token.is_cancelled()` instead of `CTRL_C` to make it cleaner and easier to maintain. If you receive `ctrl_c` and `sigterm` you could use cancelation token instead (completely removing `CTRL_C` if possible). Check https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#method.is_cancelled

Fixed! We don't even need to use `CancellationToken::is_cancelled` as `wait_or_graceful_shutdown` already races `CancellationToken::cancelled` and the future to be executed, which has the same effect.

**Review by antiyro — APPROVED** [review]
lgtm now thanks!

**Pull request merged**